src/FieldMapB.cpp

```diff
@@ -0,0 +1,377 @@
+        }
+      }
+    }
+  }
+}
+
+// get RZ cell indices corresponding to point of interest
+bool FieldMapB::GetIndices(double R, double Z, int *idxR, int *idxZ, double *deltaR, double *deltaZ)
+{
+  // boundary check
+  if( R > maxs[0] || R < mins[0] || Z > maxs[1] || Z < mins[1] ) {
+    return false;
+  }
+
+  // get indices
+  *deltaR = std::modf( (R - mins[0]) / steps[0], &ir_d );
+  *deltaZ = std::modf( (Z - mins[1]) / steps[1], &iz_d );
+
+  *idxR = static_cast<int>( ir_d );
+  *idxZ = static_cast<int>( iz_d );
+
+  return true;
+}
+
+// get XYZ cell indices corresponding to point of interest
+bool FieldMapB::GetIndices(double X, double Y, double Z, int *idxX, int *idxY, int *idxZ, double *deltaX, double *deltaY, double *deltaZ)
+{
+  // boundary check
+  if( X > maxs[0] || X < mins[0] || Y > maxs[1] || Y < mins[1] || Z > maxs[2] || Z < mins[2] ) {
+    return false;
+  }
+
+  // get indices
+  *deltaX = std::modf( (X - mins[0]) / steps[0], &ix_d );
+  *deltaY = std::modf( (Y - mins[1]) / steps[1], &iy_d );
+  *deltaZ = std::modf( (Z - mins[2]) / steps[2], &iz_d );
+
+  *idxX = static_cast<int>( ix_d );
+  *idxY = static_cast<int>( iy_d );
+  *idxZ = static_cast<int>( iz_d );
+
+  return true;
+}
+
+// load data
+void FieldMapB::LoadMap(const std::string& map_file, double scale)
+{
+  std::string   line;
+  std::ifstream input(map_file);
+  if (!input) {
+    std::cout << "FieldMapB Error: file \"" << map_file << "\" cannot be read." << std::endl;
+  }
+
+  std::vector<double> coord = {};
+  std::vector<double> Bcomp = {};
+
+  while (std::getline(input, line).good()) {
+    std::istringstream iss(line);
+
+    coord.clear();
+    Bcomp.clear();
+
+    if( case_BrBz ) {
+      coord.resize(2);
+      Bcomp.resize(2);
+      iss >> coord[0] >> coord[1] >> Bcomp[0] >> Bcomp[1];
+
+      if( ! GetIndices( coord[0], coord[1], &ir, &iz, &dr, &dz) ) {
+        std::cout << "FieldMapB Warning: coordinates out of range, skipped it."<< std::endl;
+      }
+      else {
+        Bvals_RZ[ ir ][ iz ] = { Bcomp[0] * scale, Bcomp[1] * scale };
+        //ROOT::Math::XYZPoint p(coord[0], 0, coord[1]);
+        //std::cout << p << " -> " << trans*p << std::endl;
+        //std::cout << indices[0] << ", " << indices[1] << ", " << Bcomp[0] << ", " << Bcomp[1] << std::endl;
+      }
+    }
+    else {
+      coord.resize(3);
+      Bcomp.resize(3);
+      iss >> coord[0] >> coord[1] >> coord[2] >> Bcomp[0] >> Bcomp[1] >> Bcomp[2];
+
+      if( ! GetIndices(coord[0], coord[1], coord[2], &ix, &iy, &iz, &dx, &dy, &dz) ) {
+        std::cout << "FieldMapB Warning: coordinates out of range, skipped it."<< std::endl;
+      }
+      else {
+        Bvals_XYZ[ ix ][ iy ][ iz ] = { Bcomp[0] * scale, Bcomp[1] * scale, Bcomp[2] * scale };
+      }
+    }
+  }
+}
+
+// get field components
+void FieldMapB::fieldComponents(const double* pos, double* field)
+{
+  // coordinate conversion
+  auto p = trans_inv * ROOT::Math::XYZPoint(pos[0], pos[1], pos[2]);
```

Comment:
I think trans and trans_inv can be done on reading in the map, right? We'd just apply it to the relevant mins, maxs for shifts.

Reply:
For translation transformations only, yes.  Including rotations, which is an option in the steering xml, I'm not sure.
That brings me to another question I had earlier about this line:
https://github.com/eic/epic/blob/c16103002ee38090a6ab033234b2bd544ed75349/src/FieldMapB.cpp#L267
Isn't this wrong?  We use trans_inv on the coordinates, then we use trans on the field.
Shouldn't we only have the rotation part of the transformation applied here?  Otherwise it's like translating back and forth.  In fact, it gave the wrong field values for the new BxByBz case, where I have a non-zero transformation (unlike the MARCO field).  That's why I didn't use the transformation on the BxByBz field.

Reply:
I think, it's a bug. Magnetic field vector should be only rotated. Perhaps, it was only working as long as translation was:
https://github.com/eic/epic/blob/1202217cde2bae1a0cfe521c65883bff1d035da3/compact/fields/marco.xml#L14

Reply:
That is coming from
https://github.com/eic/epic/blob/19ec3f38d53ff21b4ebb0646670a015a878927ca/src/FieldMapBrBz.cpp#L182
?

If there is a mistake, we should be seeing some azimuthal angle modulations in tracking performance.

Reply:
Removing transformation is not entirely correct. You are still supposed to apply the rotation.

Reply:
Yes, it's coming from that line.  
Br is very small in the central region, so perhaps the error on this went undetected.
Yes, we still should allow for a rotation of the field but I am unclear on how we want to do that.  From the other unresolved conversation above, do we want to rotate the field vector alone?  If so, we would have translations to move around our field region, and rotations to rotate the field vector within that translated region.

Reply:
> Br is very small in the central region, so perhaps the error on this went undetected.

Would you mind submitting a fix separately against main, so we can review and assess impact separately?

> Yes, we still should allow for a rotation of the field but I am unclear on how we want to do that. From the other unresolved conversation above, do we want to rotate the field vector alone? If so, we would have translations to move around our field region, and rotations to rotate the field vector within that translated region.

Inverse translation and inverse rotation are applied to the coordinate. Direct rotation is applied to the magnetic field. We could name those `coord_trans` and `vec_trans` respectively.

Reply:
I now have a separate translation and rotation operation in place.
I'll issue a separate fix for the potential X and Y swapping.